### ISSUE TITLE
Mismatch between total flash size and binary file

### ISSUE THREAD

**Issue by bethmal** [ebs-universe/fpv-gcc#9]
Hi,

Below 2 issues are making mismatch between total flash size and binary file.
1. For GNU arm gcc release build with -Os optimization, in .rodata section, there are place holders with specified size, but there is no allocated flash space.

Example code:
```map
 .rodata.LocalIO_X1_Get_Map_Convert.str1.1
                0x0804a857       0x8d lib\libproduct_app.a(LocalIO_Product.c.obj)
 .rodata.str1.1
                0x0804a857        0x8 lib\libproduct_app.a(LocalIO_Product.c.obj)
 *fill*         0x0804a857        0x1 
 .rodata.T1_PropTab
                0x0804a858       0x38 lib\libproduct_app.a(LocalIO_Product.c.obj)
                0x0804a858                T1_PropTab
```
2. .data section is categorized into RAM space, but it actually also stored in FLASH space.

Thanks.

**Comment by chintal** [comment]
Thank you for the issue. I will take a look and see if I can work out what's going on. 

My experience has been that different toolchains / platforms handle these sort of issues just differently enough to make it a pain to find a one size fits all solution, and I don't have the bandwidth to maintain separate dialects. I typically use this to provide a rough estimate of the totals, accurate results for specific objects, and allow for weird boundary conditions by mentally leaving margins. 

That said, I should note that it's been some while since I've seriously looked at this code. If you have a suggestion / PR for changes which does not break it for other platforms, I'll be happy to discuss it. If there is interest in the tool itself and people find it useful, I will try to make some time to restructure it to at least allow some kind of hooks to add per-platform behaviors.

**Comment by bethmal** [comment]
Hi Chintal,

Thanks for the quick response. Fully understand it's hard to support different toolchains and the rough estimation is good enough.
For issue 1, I can only think of checking symbol address as well. Besides symbol size and fill size, also track allocated actual memory size by linker.
For issue 2, .data section is stored in flash is common to all toolchains. Not sure if .data section shall be counted into both FLASH and RAM.

**Comment by chintal** [comment]
Re. 2 : .data is the section with init variables that gets copied to RAM on reset, yes? If so, then this should be counted into both Flash and RAM. If it isn't, then this is an issue. 

The only other possibility where this isn't an issue is if the map file actually holds those symbols in both flash and RAM. Additionally, we decide how to categorize each line in part based on the declared address. The flash copy and RAM copy of the stuff in .data will necessarily need to have different addresses. I don't quite remember how I handled this in fpv-gcc code, though I do remember either knowing it was correctly counted or having a reliable workaround (such as, for ex., just mentally adding up two columns). I will attempt to review some of the old map files I was testing against.

**Comment by chintal** [comment]
Re. 1: 

This type of addressing was there in most map files, yes. The total sizes should still be correct, though, since I think we do use the addresses as well. Actually, I think we use the addresses as the source of truth. Most of these symbols should have a non-zero size in the output (the size they declare), but they should show up in the DISCARDED section/column, I think. Again, I will review on some map file I have and make sure.

**Comment by bethmal** [comment]
> Re. 2 : .data is the section with init variables that gets copied to RAM on reset, yes? If so, then this should be counted into both Flash and RAM. If it isn't, then this is an issue.
> 
> The only other possibility where this isn't an issue is if the map file actually holds those symbols in both flash and RAM. Additionally, we decide how to categorize each line in part based on the declared address. The flash copy and RAM copy of the stuff in .data will necessarily need to have different addresses. I don't quite remember how I handled this in fpv-gcc code, though I do remember either knowing it was correctly counted or having a reliable workaround (such as, for ex., just mentally adding up two columns). I will attempt to review some of the old map files I was testing against.

Right, normally there is flash copy operation on reset to copy .data from FLASH to RAM.
If some map file holds .data both in both flash and RAM, that's a bit hard to recognize different style map files. Thanks for looking into this.

**Comment by bethmal** [comment]
> Re. 1:
> 
> This type of addressing was there in most map files, yes. The total sizes (on a per-file / per-build level, when seen with the sections resolved) should still be correct, though, since I think we do use the addresses as well. Actually, I think we use the addresses as the source of truth. When you drill down to the symbol level, most of these symbols should have a non-zero size in the output (the size they declare), but they should show up in the DISCARDED section/column, I think. Again, I will review on some map file I have and make sure.

Possible to track symbol allocated memory size and declared size, and present both to users?
